Python 3 compatibility issues

[flake8](http://flake8.pycqa.org) testing of https://github.com/MVIG-SJTU/AlphaPose on Python 3.7.0

$ __flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__
```
./human-detection/lib/utils/visualization.py:76:12: F821 undefined name 'xrange'
  for i in xrange(num_boxes):
           ^
./predict/json/parametric-pose-nms-MPII.py:22:14: F821 undefined name 'xrange'
    for i in xrange(len(pred_file)):
             ^
./predict/json/parametric-pose-nms-MPII.py:31:18: F821 undefined name 'xrange'
        for n in xrange(16):
                 ^
./predict/json/parametric-pose-nms-MPII.py:46:14: F821 undefined name 'xrange'
    for i in xrange(len(results)):
             ^
./predict/json/parametric-pose-nms-MPII.py:126:14: F821 undefined name 'xrange'
    for i in xrange(len(indexs)):
             ^
./predict/json/parametric-pose-nms-MPII.py:162:18: F821 undefined name 'xrange'
        for j in xrange(len(pick)):
                 ^
./predict/json/parametric-pose-nms-MPII.py:186:29: F821 undefined name 'xrange'
            for point_id in xrange(16):
                            ^
./predict/json/parametric-pose-nms-MPII.py:225:14: F821 undefined name 'xrange'
    for i in xrange(16):
             ^
./predict/json/parametric-pose-nms-COCO.py:205:33: E999 SyntaxError: invalid syntax
                   print point_id
                                ^
./train/src/misc/monitor_experiments.py:14:42: E999 SyntaxError: invalid syntax
    print "Error: No experiments provided"
                                         ^
./train/src/misc/convert_annot.py:23:14: E999 SyntaxError: invalid syntax
    print "\r",idx,
             ^
./train/src/misc/analyze_occlusion_pr.py:47:10: F821 undefined name 'xrange'
for i in xrange(thr2.size):
         ^
./train/src/misc/examples.py:51:39: E999 SyntaxError: invalid syntax
    print "Doing overall comparison..."
                                      ^
./train/src/misc/mpii.py:66:18: F821 undefined name 'xrange'
        for i in xrange(len(parts_info)):
                 ^
./train/src/misc/mpii.py:71:35: F821 undefined name 'unicode'
                    if type(v) is unicode:
                                  ^
./train/src/misc/pck_figs.py:53:40: F821 undefined name 'xrange'
        filt = np.array([True for _ in xrange(len(arr))])
                                       ^
./train/src/misc/pck_figs.py:76:14: F821 undefined name 'xrange'
    for i in xrange(len(t)):
             ^
./train/src/misc/pck_figs.py:174:10: F821 undefined name 'xrange'
for i in xrange(len(exps)):
         ^
./train/src/pypose/img.py:101:22: F821 undefined name 'xrange'
            for i in xrange(2):
                     ^
./train/src/pypose/img.py:108:14: F821 undefined name 'xrange'
    for i in xrange(3):
             ^
./train/src/pypose/img.py:117:52: F821 undefined name 'xrange'
        inp = np.array([np.fliplr(inp[i]) for i in xrange(len(inp))])
                                                   ^
./train/src/pypose/img.py:118:50: F821 undefined name 'xrange'
        hm = np.array([np.fliplr(hm[i]) for i in xrange(len(hm))])
                                                 ^
./train/src/pypose/img.py:150:14: F821 undefined name 'xrange'
    for i in xrange(3):
             ^
./train/src/pypose/eval.py:25:43: E999 SyntaxError: invalid syntax
        print "Error: Bad prediction file."
                                          ^
./train/src/pypose/report.py:549:60: E999 SyntaxError: invalid syntax
    print "This isn't too helpful, updating message soon..."
                                                           ^
./train/src/pypose/data.py:143:12: E999 SyntaxError: invalid syntax
    print ""
           ^
7     E999 SyntaxError: invalid syntax
19    F821 undefined name 'xrange'
26
```

--- a/human-detection/lib/utils/visualization.py
+++ b/human-detection/lib/utils/visualization.py
@@ -8,6 +8,7 @@ from __future__ import division
 from __future__ import print_function
 
 import numpy as np
+from six.moves import xrange
 import PIL.Image as Image
 import PIL.ImageColor as ImageColor
 import PIL.ImageDraw as ImageDraw

--- a/predict/json/parametric-pose-nms-COCO.py
+++ b/predict/json/parametric-pose-nms-COCO.py
@@ -4,9 +4,11 @@ Created on Wed Nov  9 10:58:09 2016
 
 @author: benjamin
 """
+from __future__ import print_function
 
 import json
 import numpy as np
+from six.moves import xrange
 import h5py
 import os
 import zipfile
@@ -202,7 +204,7 @@ def test_parametric_pose_NMS_json(delta1,delta2,mu,gamma,outputpath):
             for point_id in xrange(17):
                 if math.isnan(merge_poses[point_id,0]):
                    merge_poses[point_id,0] = 0
-                   print point_id
+                   print(point_id)
                 if math.isnan(merge_poses[point_id,1]):
                    merge_poses[point_id,1] = 1
                 NMS_preds.write("\t{}\t{}".format(merge_poses[point_id,0]-0.25,merge_poses[point_id,1]-0.25))

--- a/predict/json/parametric-pose-nms-MPII.py
+++ b/predict/json/parametric-pose-nms-MPII.py
@@ -7,6 +7,7 @@ Created on Wed Nov  9 10:58:09 2016
 
 import json
 import numpy as np
+from six.moves import xrange
 import h5py
 import os
 import argparse

--- a/train/src/misc/analyze_occlusion_pr.py
+++ b/train/src/misc/analyze_occlusion_pr.py
@@ -3,6 +3,7 @@ import matplotlib.pyplot as plt
 from pypose import ref
 import h5py
 import numpy as np
+from six.moves import xrange
 
 a = ref.load('mpii','valid')
 

--- a/train/src/misc/convert_annot.py
+++ b/train/src/misc/convert_annot.py
@@ -1,5 +1,7 @@
+from __future__ import print_function
 import h5py
 import numpy as np
+from six.moves import xrange
 import sys
 import mpii
 
@@ -20,7 +22,7 @@ with h5py.File('../../data/mpii/annot/multi-idxs.h5','r') as f:
 imgnameRef = mpii.annot['annolist'][0][0][0]['image'][:]
 
 for idx in xrange(mpii.nimages):
-    print "\r",idx,
+    print("\r",idx, end=' ')
     sys.stdout.flush()
 
     for person in xrange(mpii.numpeople(idx)):
@@ -58,7 +60,7 @@ for idx in xrange(mpii.nimages):
                 else:   # Training image (something missing in annot)
                     annot['istrain'] += [2]
 
-print ""
+print("")
 
 with h5py.File('mpii-annot.h5','w') as f:
     f.attrs['name'] = 'mpii'

--- a/train/src/misc/examples.py
+++ b/train/src/misc/examples.py
@@ -1,12 +1,13 @@
+from __future__ import print_function
 # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 # All of these examples are really, really outdated but offer some insights 
 # into using the python code, if you want to check it out
 # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-
 import numpy as np
 import pypose as pose
 import pypose.mpii as ds  # Use this to swap which dataset you want to use
+from six.moves import xrange
 
 # Sample dataset generation
 if False:
@@ -48,10 +49,10 @@ if False:
     pdf = pose.report.PdfPages(pose.ref.posedir+'/img/reports/fwd_back_sample.pdf')
 
     # Add whatever pages you want
-    print "Doing overall comparison..."
+    print("Doing overall comparison...")
     pose.report.filtercomparison(ds.name, dists, filts, filtnames=filtnames, title=title, pdf=pdf)
     for i,filt in enumerate(filts[:-1]):
-        print "Generating images for - %s..." % filtnames[i]
+        print("Generating images for - %s..." % filtnames[i])
         pose.report.sampleimages(ds, preds, dists=dists, pdf=pdf, title=filtnames[i], filt=filt)
         pose.report.sampleimages(ds, preds, dists=dists, pdf=pdf, title=filtnames[i], filt=filt, get_worst=True)
 
@@ -81,11 +82,11 @@ if True:
     title='Performance Comparison - Torso Deviation from Vertical'
     pdf = pose.report.PdfPages(pose.ref.posedir+'/img/reports/torso_angle_sample.pdf')
 
-    print "Doing overall comparison..."
+    print("Doing overall comparison...")
     pose.report.filtercomparison(ds.name, dists, filts, filtnames=filtnames, title=title, pdf=pdf)
     for i in xrange(7):
         # This loop will only generate poor performing images for the first filter (people who are upright)
-        print "Generating images for page - %d..." % i
+        print("Generating images for page - %d..." % i)
         pose.report.sampleimages(ds, preds, dists=dists, pdf=pdf, title=filtnames[0], filt=filts[0],
                                 get_worst=True, page_num=i+1)
 

--- a/train/src/misc/monitor_experiments.py
+++ b/train/src/misc/monitor_experiments.py
@@ -1,3 +1,4 @@
+from __future__ import print_function
 import numpy as np
 import matplotlib
 import matplotlib.pyplot as plt
@@ -11,17 +12,17 @@ plt.ion()
 try:
     experiments_to_show = sys.argv[1].split(',')
 except:
-    print "Error: No experiments provided"
+    print("Error: No experiments provided")
     exit()
 
-print "Monitoring the following experiments:",
-for exp in experiments_to_show: print exp,
-print ""
+print("Monitoring the following experiments:", end=' ')
+for exp in experiments_to_show: print(exp, end=' ')
+print("")
 
 track_multiple = sys.argv[2] == '1'
 if track_multiple:
     exp_to_track = experiments_to_show[0]
-    print "Tracking all variations of:",exp_to_track
+    print("Tracking all variations of:",exp_to_track)
     experiments_to_show.remove(exp_to_track)
 
 def readlog(filepath):
@@ -55,9 +56,9 @@ while True:
                     if not subdirname in experiments_to_show:
                         experiments_to_show += [subdirname]
 
-    print "Updated experiments to show:",
-    for exp in experiments_to_show: print exp,
-    print ""
+    print("Updated experiments to show:", end=' ')
+    for exp in experiments_to_show: print(exp, end=' ')
+    print("")
 
     idx = [1, 2, 0] # Epoch, Loss, Accuracy indices
 
@@ -111,6 +112,6 @@ while True:
             axs[k].set_ylim(0,1)
 
     axs['Test accuracy'].legend(loc='lower right', fontsize=10)
-    print time.strftime('%X %x %Z')
+    print(time.strftime('%X %x %Z'))
     plt.show()
     plt.pause(900)

--- a/train/src/misc/mpii.py
+++ b/train/src/misc/mpii.py
@@ -1,5 +1,7 @@
 import scipy.io
 import scipy.misc
+from six import text_type
+from six.moves import xrange
 import numpy as np
 
 # Load in annotations
@@ -68,7 +70,7 @@ def partinfo(idx, person, part):
                 if 'is_visible' in parts_info.dtype.fields:
                     v = parts_info[i]['is_visible']
                     v = v[0][0] if len(v) > 0 else 1
-                    if type(v) is unicode:
+                    if type(v) is text_type:
                         v = int(v)
                 else:
                     v = 1

--- a/train/src/misc/pck_figs.py
+++ b/train/src/misc/pck_figs.py
@@ -1,6 +1,7 @@
 import matplotlib.pyplot as plt
 from scipy.io import loadmat
 import numpy as np
+from six.moves import xrange
 import h5py
 
 def setuppdjplot(ax, i):

--- a/train/src/pypose/data.py
+++ b/train/src/pypose/data.py
@@ -1,5 +1,7 @@
+from __future__ import print_function
 import numpy as np
 import h5py
+from six.moves import xrange
 import getopt, sys, time
 
 import img
@@ -77,15 +79,15 @@ def gendetect(annot, idx, img_in, chg=None):
         #     r = np.random.randint(-chg['rotate'], chg['rotate'] + 1)
         # Translation
         if 'translate' in chg:
-            for i in xrange(2):
+            for i in range(2):
                 offset = np.random.randint(-chg['translate'], chg['translate'] + 1)
-                c[i] += offset
+                img_c[i] += offset
 
     img_c[0] += img_s * np.random.randint(-10,10)
     img_c[1] += img_s * np.random.randint(-10,10)
     cropped = img.crop(img_in, img_c, img_s, ref.in_res)
     inp = np.zeros((3, ref.in_res[0], ref.in_res[1]))
-    for i in xrange(3): inp[i, :, :] = cropped[:, :, i]
+    for i in range(3): inp[i, :, :] = cropped[:, :, i]
 
     out = np.zeros((2, ref.out_res[0], ref.out_res[1]))
     for i in idxs:
@@ -140,11 +142,11 @@ def generateset(dataset, settype, filename, numsamples, datadir=None, chg=None, 
         numsamples = numavailable
         inorder = True
 
-    print ""
-    print "Generating %s %s set: %s" % (dataset, sampletype, settype)
-    print "Path to dataset: %s" % filepath
-    print "Number of samples: %d" % numsamples
-    print "Data augmentation: %s" % (str(chg))
+    print("")
+    print("Generating %s %s set: %s" % (dataset, sampletype, settype))
+    print("Path to dataset: %s" % filepath)
+    print("Number of samples: %d" % numsamples)
+    print("Data augmentation: %s" % (str(chg)))
 
     # Data/label sizes can be all over the place, this is the easiest way to check
     ex_in, ex_out = gensample(annot, 0, chg=chg, sampletype=sampletype, extra_args=extra_args)
@@ -155,9 +157,9 @@ def generateset(dataset, settype, filename, numsamples, datadir=None, chg=None, 
     ref_idxs = np.zeros((numsamples, 1), np.float32)
 
     # Loop to generate new samples
-    print ''
-    print '| Progress            |'
-    print '|',
+    print('')
+    print('| Progress            |')
+    print('|', end=' ')
     sys.stdout.flush()
 
     starttime = time.time()
@@ -170,14 +172,14 @@ def generateset(dataset, settype, filename, numsamples, datadir=None, chg=None, 
         ref_idxs[i] = idx
 
         if i % (numsamples/10) == 0:
-            print '=',
+            print('=', end=' ')
             sys.stdout.flush()
 
-    print '|'
-    print ''
-    print 'Done!',
-    print '(%.2f seconds to complete.)' % (time.time() - starttime)
-    print ''
+    print('|')
+    print('')
+    print('Done!', end=' ')
+    print('(%.2f seconds to complete.)' % (time.time() - starttime))
+    print('')
 
     # Write out to hdf5 files
     with h5py.File(filepath, 'w') as f:
@@ -186,29 +188,29 @@ def generateset(dataset, settype, filename, numsamples, datadir=None, chg=None, 
         f['index'] = ref_idxs
 
 def helpmessage():
-    print "Extra flags:"
-    print "  -d, --dataset    :: Datset choice (mpii or flic), REQUIRED"
-    print "  -o, --outfile    :: Output file for data (do not include '.h5'), REQUIRED"
-    print "  -p, --prefix     :: Directory to save to (no need to include dataset name)"
-    print "  -t, --type       :: Dataset type (train or test), default is train"
-    print "  -n, --numsamples :: Number of samples to generate, default is all available (-1) for test and 100 for train"
-    print ""
-    print "Augmentation options: (default Tompson's options for train, none for test)"
-    print "  -m, --move       :: Translate (0 - 50)"
-    print "  -z, --zoom       :: Scale (0.0 - 1.0)"
-    print "  -r, --rotate     :: Rotate (-1 for fixed vertical, 0-180 for max distortion)"
-    print "  (Tompson's options are: -m 0 -z .5 -r 20"
-    print ""
-    print "Other dataset types:"
-    print "  -q, --detect"
-    print "  -c, --cascade     :: Provide first joint as argument, must use additional argument below"
-    print "  -j, --pairedjoint :: Provide second joint to be used with 'cascade'"
-    print ""
-    print "Additional limb heatmap output:"
-    print "  -s, --segment    :: - 0 No limb segment output (default)"
-    print "                      - 1 Does not distinguish parts, angle == angle + 180"
-    print "                      - 2 Distinguishes different part types, angle == angle + 180"
-    print "                      - 3 Distinguishes different part types, angle != angle + 180"
+    print("Extra flags:")
+    print("  -d, --dataset    :: Datset choice (mpii or flic), REQUIRED")
+    print("  -o, --outfile    :: Output file for data (do not include '.h5'), REQUIRED")
+    print("  -p, --prefix     :: Directory to save to (no need to include dataset name)")
+    print("  -t, --type       :: Dataset type (train or test), default is train")
+    print("  -n, --numsamples :: Number of samples to generate, default is all available (-1) for test and 100 for train")
+    print("")
+    print("Augmentation options: (default Tompson's options for train, none for test)")
+    print("  -m, --move       :: Translate (0 - 50)")
+    print("  -z, --zoom       :: Scale (0.0 - 1.0)")
+    print("  -r, --rotate     :: Rotate (-1 for fixed vertical, 0-180 for max distortion)")
+    print("  (Tompson's options are: -m 0 -z .5 -r 20")
+    print("")
+    print("Other dataset types:")
+    print("  -q, --detect")
+    print("  -c, --cascade     :: Provide first joint as argument, must use additional argument below")
+    print("  -j, --pairedjoint :: Provide second joint to be used with 'cascade'")
+    print("")
+    print("Additional limb heatmap output:")
+    print("  -s, --segment    :: - 0 No limb segment output (default)")
+    print("                      - 1 Does not distinguish parts, angle == angle + 180")
+    print("                      - 2 Distinguishes different part types, angle == angle + 180")
+    print("                      - 3 Distinguishes different part types, angle != angle + 180")
     sys.exit(2)
 
 def main(argv):
@@ -230,7 +232,7 @@ def main(argv):
                                                                    "numsamples=", "move=", "zoom=", "rotate=",
                                                                    "segment=", "detect", "cascade=", "pairedjoint="])
     except getopt.GetoptError:
-        print "Incorrect arguments"
+        print("Incorrect arguments")
         helpmessage()
         sys.exit()
     for opt,arg in opts:
@@ -241,7 +243,7 @@ def main(argv):
         elif opt in ('-d','--dataset'):
             dataset = arg
             if not (dataset in ['mpii', 'flic']):
-                print "Bad argument for --dataset"
+                print("Bad argument for --dataset")
                 helpmessage()
         # Output file
         elif opt in ('-o','--outfile'):
@@ -253,19 +255,19 @@ def main(argv):
         elif opt in ('-t','--type'):
             settype = arg
             if not (settype in ['train','test','valid','train_obs','test_obs']):
-                print "Bad argument for --type"
+                print("Bad argument for --type")
                 helpmessage()
         # Number of samples
         elif opt in ('-n','--numsamples'):
             numsamples = int(arg)
             if numsamples < -1:
-                print "Bad argument for --numsamples"
+                print("Bad argument for --numsamples")
                 helpmessage()
         # Move
         elif opt in ('-m','--move'):
             move = int(arg)
             if not 0 <= move <= 50:
-                print "Bad argument for --move"
+                print("Bad argument for --move")
                 helpmessage()
             else:
                 if chg is None:
@@ -275,7 +277,7 @@ def main(argv):
         elif opt in ('-z','--zoom'):
             zoom = float(arg)
             if not 0 <= zoom <= 1:
-                print "Bad argument for --zoom"
+                print("Bad argument for --zoom")
                 helpmessage()
             else:
                 if chg is None:
@@ -285,7 +287,7 @@ def main(argv):
         elif opt in ('-r','--rotate'):
             rot = int(arg)
             if not -1 <= rot <= 180:
-                print "Bad argument for --rotate"
+                print("Bad argument for --rotate")
                 helpmessage()
             else:
                 if chg is None:
@@ -295,7 +297,7 @@ def main(argv):
         elif opt in ('-s','--segment'):
             seg = int(arg)
             if not (0 <= seg <= 3):
-                print "Bad argument for --segment"
+                print("Bad argument for --segment")
                 helpmessage()
         # Detect
         elif opt in ('-q','--detect'):
@@ -308,10 +310,10 @@ def main(argv):
             jnt2 = int(arg)
 
     if dataset is None:
-        print "No dataset chosen."
+        print("No dataset chosen.")
         helpmessage()
     if outfile is None:
-        print "No output filename chosen."
+        print("No output filename chosen.")
         helpmessage()
 
     if settype in ['test','test_obs']:
@@ -326,7 +328,7 @@ def main(argv):
     # If we're generating cascade data make sure two joints have been provided
     if sampletype == 'cascade':
         if jnt1 == -1 or jnt2 == -1:
-            print "Need two joints to generate cascade data"
+            print("Need two joints to generate cascade data")
             helpmessage()
         extra = [jnt1, jnt2]
 

--- a/train/src/pypose/eval.py
+++ b/train/src/pypose/eval.py
@@ -1,4 +1,6 @@
+from __future__ import print_function
 import numpy as np
+from six.moves import xrange
 import ref
 import img
 
@@ -22,7 +24,7 @@ def getdists(pred, dotrain=False):
     elif pred.shape[1] == 16:
         dataset = 'mpii'
     else:
-        print "Error: Bad prediction file."
+        print("Error: Bad prediction file.")
         return 0
 
     idx_ref = []

--- a/train/src/pypose/img.py
+++ b/train/src/pypose/img.py
@@ -1,10 +1,12 @@
 import numpy as np
 import scipy.misc
 import scipy.signal
+from six.moves import xrange
 import math
 
 import draw
 import ref
+
 
 # =============================================================================
 # General image processing functions

--- a/train/src/pypose/report.py
+++ b/train/src/pypose/report.py
@@ -1,8 +1,10 @@
+from __future__ import print_function
 import numpy as np
 import matplotlib
 matplotlib.use('Agg')
 import matplotlib.pyplot as plt
 from matplotlib.backends.backend_pdf import PdfPages
+from six.moves import xrange
 import sys, getopt
 import h5py
 
@@ -546,7 +548,7 @@ def limbreport(dataset, exp_id, pdf=None):
 # =============================================================================
 
 def helpmessage():
-    print "This isn't too helpful, updating message soon..."
+    print("This isn't too helpful, updating message soon...")
     sys.exit(2)
 
 # Main
@@ -566,7 +568,7 @@ def main(argv):
         opts, args = getopt.getopt(argv, "hd:e:c:p:viol", ["help", "dataset=", "expID=", "compare=", "prev=",
                                                            "vert", "images", "obs", "limb"])
     except getopt.GetoptError:
-        print "Incorrect arguments"
+        print("Incorrect arguments")
         helpmessage()
         sys.exit(2)
     for opt, arg in opts:
@@ -582,13 +584,13 @@ def main(argv):
         elif opt in ('-l', '--limb'): limb = True
 
     if dataset is None:
-        print "No dataset chosen."
+        print("No dataset chosen.")
         helpmessage()
     if not (dataset in ['mpii','flic']):
-        print "Bad argument for --dataset"
+        print("Bad argument for --dataset")
         helpmessage()
     if exp_id is None:
-        print "No experiment number provided."
+        print("No experiment number provided.")
         helpmessage()
     expdir = ref.posedir + '/exp/' + dataset + '/' + exp_id
 
@@ -608,7 +610,7 @@ def main(argv):
         annot = ref.load(dataset, 'valid')
 
     # Load predictions
-    print "Loading predictions"
+    print("Loading predictions")
     pred_opts = [num_parts, vert, obs]
     preds, dists, _ = loadpreds(dataset, expdir + '/preds.h5', pred_opts, images)
 
@@ -646,7 +648,7 @@ def main(argv):
 
     filt = None
 
-    print "Creating overview page"
+    print("Creating overview page")
     # Main report creation
     pdf = PdfPages(expdir + '/report.pdf')
 
@@ -655,7 +657,7 @@ def main(argv):
                      other_dists=other_dists, pdf=pdf)
 
     if images:
-        print "Creating prediction examples page"
+        print("Creating prediction examples page")
         # Overall performance examples
         num_good_exs = 2
         num_bad_exs = 6

